core/src/test/java/org/elasticsearch/index/query/BaseQueryTestCase.java

```diff
@@ -311,4 +298,30 @@ protected static void assertValidate(QueryBuilder queryBuilder, int totalExpecte
             assertThat(queryValidationException, nullValue());
         }
     }
-}
+
+    protected String[] getRandomTypes() {
+        String[] types;
+        if (currentTypes.length > 0 && randomBoolean()) {
+            int numberOfQueryTypes = randomIntBetween(1, currentTypes.length);
+            types = new String[numberOfQueryTypes];
+            for (int i = 0; i < numberOfQueryTypes; i++) {
+                types[i] = randomFrom(currentTypes);
+            }
+        } else {
+            if (randomBoolean()) {
+                types = new String[] { MetaData.ALL };
+            } else {
+                types = new String[0];
+            }
+        }
+        return types;
+    }
+
+    protected String getRandomType() {
+        String[] randomTypes = getRandomTypes();
+        if (randomTypes.length == 0) {
+            return MetaData.ALL;
+        }
+        return randomFrom(randomTypes);
```

Comment:
then handle when currentTypes is empty and make it all, but don't call getRandomTypes for nothing. Why would you need to create an array composed of a random number of types out of the current types, to then keep only one. Just keep one from original array? Or am I missing something?